webapp/src/js/services/contact-save.js

```diff
@@ -62,113 +62,103 @@ angular.module('inboxServices').service('ContactSave',
       }
 
       return doc;
-    }
+    };
 
-    function prepareRepeatedDocs(doc, repeated) {
-      if (repeated && repeated.child_data) {
-        return _.map(repeated.child_data, function(child) {
-          child.parent = ExtractLineage(doc);
-          return prepare(child);
-        });
-      } else {
-        return [];
-      }
-    }
+    const prepareRepeatedDocs = (doc, repeated) => {
+      const childData = (repeated && repeated.child_data) || [];
+      return childData.map(child => {
+        child.parent = ExtractLineage(doc);
+        return prepare(child);
+      });
+    };
 
-    function extractIfRequired(name, value) {
-      if (name === 'parent' || name === 'contact') {
-        return ExtractLineage(value);
-      }
-      return value;
-    }
+    const extractIfRequired = (name, value) => {
+      return CONTACT_FIELD_NAMES.includes(name) ? ExtractLineage(value) : value;
+    };
 
     // Mutates the passed doc to attach prepared siblings, and returns all
     // prepared siblings to be persisted.
     // This will (on a correctly configured form) attach the full parent to
     // doc, and in turn siblings. See internal comments.
-    function prepareAndAttachSiblingDocs(schema, doc, original, siblings) {
+    const prepareAndAttachSiblingDocs = (doc, original, siblings) => {
       if (!doc._id) {
         throw new Error('doc passed must already be prepared with an _id');
       }
 
-      var preparedSiblings = [];
-      var promiseChain = $q.resolve();
+      const preparedSiblings = [];
+      let promiseChain = $q.resolve();
 
-      _.each(schema.fields, function(fieldConf, fieldName) {
-        var customType = fieldConf.type.match(/^(db|custom):(.*)$/);
-        if (customType) {
-          var value = doc[fieldName];
-          if (_.isObject(value)) {
-            value = doc[fieldName]._id;
+      CONTACT_FIELD_NAMES.forEach(fieldName => {
+        let value = doc[fieldName];
+        if (_.isObject(value)) {
+          value = doc[fieldName]._id;
+        }
+        if (!value) {
+          return;
+        }
+        if (value === 'NEW') {
+          const preparedSibling = prepare(siblings[fieldName]);
+
+          // by default all siblings are "person" types but can be overridden
+          // by specifying the type and contact_type in the form
+          if (!preparedSibling.type) {
+            preparedSibling.type = 'person';
```

Comment:
We're not defaulting to using `contact_type` and `type` for new contacts?

Reply:
Initially I did, but then I realised that could break existing configurations (tasks for example) which may check the `type` field, so I decided to be backwards compatible here.

This can go away in the next major and we can require everyone to check their config...

Reply:
Thanks for explaining!